locations/spiders/gamestop_us.py

```diff
@@ -0,0 +1,97 @@
+import json
+
+from scrapy import Spider
+from scrapy.http import FormRequest
+
+from locations.categories import Categories
+from locations.dict_parser import DictParser
+from locations.geo import postal_regions
+from locations.hours import OpeningHours
+from locations.user_agents import BROWSER_DEFAULT
+
+GAMESTOP_SHARED_ATTRIBUTES = {
+    "brand": "GameStop",
+    "brand_wikidata": "Q202210",
+    "extras": Categories.SHOP_VIDEO_GAMES.value,
+}
+
+
+class GamestopUSSpider(Spider):
+    name = "gamestop_us"
+    item_attributes = GAMESTOP_SHARED_ATTRIBUTES
+    allowed_domains = ["www.gamestop.com"]
+    start_urls = [
+        "https://www.gamestop.com/on/demandware.store/Sites-gamestop-us-Site/default/Stores-FindStores?hasCondition=false&hasVariantsAvailableForLookup=false&hasVariantsAvailableForPickup=false&source=plp&showMap=false&products=undefined:1"
+    ]
+    user_agent = BROWSER_DEFAULT
+    custom_settings = {"ROBOTSTXT_OBEY": False}
+    download_delay = 0.2
+
+    def start_requests(self):
+        headers = {
+            "Accept": "application/json",
+            "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
+            "Referer": "https://www.gamestop.com/stores/",
+            "X-Requested-With": "XMLHttpRequest",
+        }
+        for url in self.start_urls:
+            # There appears to be no way via the website or API to
+            # list all stores or perform a coordinate/radius or
+            # bounding box search. A postcode search appears to be
+            # the only option and there are almost 38000 postcodes
+            # in the US (38000 API requests to Gamestop). It is
+            # possible to reduce the list of postcodes to search
+            # by ignoring postcodes with low populations. There are
+            # fancier and more accurate ways to reduce the list of
+            # postcodes to those which are useful, but this is not
+            # yet implemented in ATP and is a complex problem
+            # requiring significant experimentation and
+            # documentation to explain the methodology which others
+            # can reproduce.
+            #
+            # The 2022 Annual Report (https://news.gamestop.com/static-files/fe325562-c087-4fad-809c-efd183364196)
+            # states that there were 2949 open stores at 28 Jan 2023
+            # operating in the US.
+            #
+            # One year later, the following search results were
```

Comment:
The API looks to be limited by size, 3MB and 1 million chars, not the radius. Pushing the radius up high quickly hits the limit, however 200 radius and 50000 population gives me:

```
 'downloader/response_count': 1017,
 'downloader/response_status_count/200': 1017,
 'item_scraped_count': 2824,
```

Reply:
Thanks for finding that the radius can be more than 100! And for confirming a population filter of 50000 gives good results :)

I went on a wild hunt for algorithms but found nothing suitable for preparing a minimal set of postcode centroids buffered by a radius, where the resulting circles fully cover the USA (or at least populated areas of the USA). "Point thinning", "Point decimation", and similar approaches were giving me nonsensical results where it'd reduce the number of points, but still keep clusters of 10 or so points all within the same close proximity (no need to keep all 10).

I also implemented a "consolidate_cities" feature in this latest patch that gets close to halving the number of API requests required without impacting results:

30000 population at 200 radius with city consolidation
```
 'downloader/response_count': 2013,
 'downloader/response_status_count/200': 2013,
 'item_scraped_count': 2843,
```

50000 population at 200 radius with city consolidation
```
 'downloader/response_count': 629,
 'downloader/response_status_count/200': 629,
 'item_scraped_count': 2824,
```